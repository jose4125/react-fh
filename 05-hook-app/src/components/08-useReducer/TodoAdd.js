import { useForm } from "../../hooks/useForm";

export const TodoAdd = ({ handleAdd }) => {
  const { formState, handleInputChange, resetForm } = useForm({
    desc: "",
  });

  const handleSubmit = (event) => {
    event.preventDefault();

    if (!formState.desc.trim()) return;

    const newTodo = {
      id: new Date().getTime(),
      desc: formState.desc,
      done: false,
    };

    handleAdd(newTodo);
    resetForm();
  };

  return (
    <>
      <h4>agregar todo</h4>
      <span>-----------------------------------------------</span>
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="desc"
          placeholder="nuevo todo"
          autoComplete="off"
          value={formState.desc}
          onChange={handleInputChange}
        />
        <button type="submit">agregar</button>
      </form>
    </>
  );
};
