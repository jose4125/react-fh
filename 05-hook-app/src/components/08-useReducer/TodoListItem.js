export const TodoListItem = ({ todo, index, handleComplete, handleRemove }) => {
  const { id, desc, done } = todo;

  return (
    <li>
      <hr />
      <p onClick={() => handleComplete(id)}>
        {index + 1} {desc}
        {done && " (done, yeah!!)"}
      </p>
      <button onClick={() => handleRemove(id)}>borrar</button>
      <hr />
    </li>
  );
};
