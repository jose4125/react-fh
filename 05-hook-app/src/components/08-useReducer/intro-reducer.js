const initialState = [
  {
    id: 1,
    todo: "comprar pan",
    done: false,
  },
];

const todoReducer = (state = initialState, action) => {
  switch (action?.type) {
    case "ADD_TODO":
      return [...state, action.payload];
    default:
      return state;
  }
};

let todos = todoReducer();

const newTodo = {
  id: 2,
  todo: "estudiar",
  done: false,
};

const addAction = {
  type: "ADD_TODO",
  payload: newTodo,
};

todos = todoReducer(todos, addAction);
console.log(todos);
