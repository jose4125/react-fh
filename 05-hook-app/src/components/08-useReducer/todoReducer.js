export const todoReducer = (state = [], action) => {
  switch (action?.type) {
    case "ADD_TODO":
      return [action.payload, ...state];
    case "REMOVE_TODO":
      return state.filter((todo) => todo.id !== action.payload);
    case "COMPLETE_TODO":
      return state.map(
        (todo) =>
          todo.id === action.payload ? { ...todo, done: !todo.done } : todo
      );
    default:
      return state;
  }
};
