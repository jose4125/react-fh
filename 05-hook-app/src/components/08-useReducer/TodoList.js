import { TodoListItem } from "./TodoListItem";

export const TodoList = ({ todos, ...props }) => {
  return (
    <ul>
      {todos.map((todo, index) => {
        return (
          <TodoListItem key={todo.id} todo={todo} index={index} {...props} />
        );
      })}
    </ul>
  );
};
