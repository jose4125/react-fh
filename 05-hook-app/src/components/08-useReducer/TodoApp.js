import { useEffect, useReducer } from "react";
import { todoReducer } from "./todoReducer";
import { TodoAdd } from "./TodoAdd";
import { TodoList } from "./TodoList";

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) ?? [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(
    () => {
      localStorage.setItem("todos", JSON.stringify(todos));
    },
    [todos]
  );

  const handleRemove = (todoId) => {
    const action = {
      type: "REMOVE_TODO",
      payload: todoId,
    };

    dispatch(action);
  };

  const handleComplete = (todoId) => {
    const action = {
      type: "COMPLETE_TODO",
      payload: todoId,
    };
    dispatch(action);
  };

  const handleAdd = (newTodo) => {
    dispatch({
      type: "ADD_TODO",
      payload: newTodo,
    });
  };

  const todoListProps = {
    todos: todos,
    handleRemove: handleRemove,
    handleComplete: handleComplete,
  };

  return (
    <div>
      <h1>Todo app ({todos.length})</h1>
      <hr />
      <div>
        <TodoAdd handleAdd={handleAdd} />
      </div>
      <TodoList {...todoListProps} />
    </div>
  );
};
