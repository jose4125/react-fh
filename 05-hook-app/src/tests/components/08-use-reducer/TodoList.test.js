import { shallow } from "enzyme";
import { TodoList } from "../../../components/08-useReducer/TodoList";
import { demoTodos } from "../../fixtures/demoTodos";

describe("test TodoList", () => {
  let wrapper;
  const props = {
    todos: demoTodos,
    handleComplete: jest.fn(),
    handleRemove: jest.fn(),
  };

  beforeEach(() => {
    wrapper = shallow(<TodoList {...props} />);
  });

  test("should render properly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Should have TodoListItem", () => {
    const todoList = wrapper.find("TodoListItem");

    expect(todoList).toHaveLength(demoTodos.length);
  });

  test("Should have the correct props", () => {
    const todo = wrapper.find("TodoListItem").at(0);

    expect(todo.prop("todo")).toEqual(demoTodos[0]);
    expect(todo.prop("index")).toBe(0);
    expect(typeof todo.prop("handleRemove")).toBe("function");
    expect(typeof todo.prop("handleComplete")).toBe("function");
  });
});
