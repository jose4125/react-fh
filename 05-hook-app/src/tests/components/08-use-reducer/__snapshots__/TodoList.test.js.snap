// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test TodoList should render properly 1`] = `
<ul>
  <TodoListItem
    handleComplete={[MockFunction]}
    handleRemove={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "aprender react",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleComplete={[MockFunction]}
    handleRemove={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "aprender node",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
